panic: runtime error: invalid memory address or nil pointer dereference

#### Overview of the Issue

(consul_1.2.2) Version: 'v1.2.2'
==> Starting Consul agent...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x106b896]

goroutine 1 [running]:
github.com/hashicorp/consul/agent.(*Agent).consulConfig(0xc42012e000, 0xc42009a2d0, 0xdf8475800, 0xc420094240)
	/go/src/github.com/hashicorp/consul/agent/agent.go:824 +0x3d6
github.com/hashicorp/consul/agent.(*Agent).Start(0xc42012e000, 0xc42012e000, 0x0)
	/go/src/github.com/hashicorp/consul/agent/agent.go:311 +0x2a7
github.com/hashicorp/consul/command/agent.(*cmd).run(0xc420259800, 0xc42002e0b0, 0x1, 0x1, 0x0)
	/go/src/github.com/hashicorp/consul/command/agent/agent.go:219 +0x454
github.com/hashicorp/consul/command/agent.(*cmd).Run(0xc420259800, 0xc42002e0b0, 0x1, 0x1, 0xc42038ef00)
	/go/src/github.com/hashicorp/consul/command/agent/agent.go:74 +0x50
github.com/hashicorp/consul/vendor/github.com/mitchellh/cli.(*CLI).Run(0xc4201ff320, 0xc4201ff320, 0x40, 0xc42038f200)
	/go/src/github.com/hashicorp/consul/vendor/github.com/mitchellh/cli/cli.go:242 +0x1eb
main.realMain(0xc420094058)
	/go/src/github.com/hashicorp/consul/main.go:53 +0x3ee
main.main()
	/go/src/github.com/hashicorp/consul/main.go:20 +0x22

#### Overview of the Issue
(consul_1.0.0) Version: 'v1.0.0'
==> Starting Consul agent...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xe53d4e]

goroutine 1 [running]:
github.com/hashicorp/consul/agent.(*Agent).consulConfig(0xc42021c4e0, 0xc420158320, 0xc42015e360, 0xc420178d80)
	/gopath/src/github.com/hashicorp/consul/agent/agent.go:646 +0x32e
github.com/hashicorp/consul/agent.(*Agent).Start(0xc42021c4e0, 0xc42021c4e0, 0x0)
	/gopath/src/github.com/hashicorp/consul/agent/agent.go:248 +0x169
github.com/hashicorp/consul/command.(*AgentCommand).run(0xc4202bc1e0, 0xc42000e0b0, 0x1, 0x1, 0x0)
	/gopath/src/github.com/hashicorp/consul/command/agent.go:320 +0x4e0
github.com/hashicorp/consul/command.(*AgentCommand).Run(0xc4202bc1e0, 0xc42000e0b0, 0x1, 0x1, 0xc4201d3620)
	/gopath/src/github.com/hashicorp/consul/command/agent.go:267 +0x50
github.com/hashicorp/consul/vendor/github.com/mitchellh/cli.(*CLI).Run(0xc420229320, 0xc420229320, 0x40, 0xc42029f6a0)
	/gopath/src/github.com/hashicorp/consul/vendor/github.com/mitchellh/cli/cli.go:242 +0x1eb
main.realMain(0xef5bd7)
	/gopath/src/github.com/hashicorp/consul/main.go:52 +0x3fb
main.main()
	/gopath/src/github.com/hashicorp/consul/main.go:19 +0x22

--- a/agent/agent.go
+++ b/agent/agent.go
@@ -788,6 +788,9 @@ func (a *Agent) consulConfig() (*consul.Config, error) {
 	base.SerfLANConfig.MemberlistConfig.ProbeTimeout = a.config.GossipLANProbeTimeout
 	base.SerfLANConfig.MemberlistConfig.SuspicionMult = a.config.GossipLANSuspicionMult
 	base.SerfLANConfig.MemberlistConfig.RetransmitMult = a.config.GossipLANRetransmitMult
+	if a.config.ReconnectTimeoutLAN != 0 {
+		base.SerfLANConfig.ReconnectTimeout = a.config.ReconnectTimeoutLAN
+	}
 
 	if a.config.SerfBindAddrWAN != nil {
 		base.SerfWANConfig.MemberlistConfig.BindAddr = a.config.SerfBindAddrWAN.IP.String()
@@ -802,6 +805,9 @@ func (a *Agent) consulConfig() (*consul.Config, error) {
 		base.SerfWANConfig.MemberlistConfig.ProbeTimeout = a.config.GossipWANProbeTimeout
 		base.SerfWANConfig.MemberlistConfig.SuspicionMult = a.config.GossipWANSuspicionMult
 		base.SerfWANConfig.MemberlistConfig.RetransmitMult = a.config.GossipWANRetransmitMult
+		if a.config.ReconnectTimeoutWAN != 0 {
+			base.SerfWANConfig.ReconnectTimeout = a.config.ReconnectTimeoutWAN
+		}
 	} else {
 		// Disable serf WAN federation
 		base.SerfWANConfig = nil
@@ -809,13 +815,6 @@ func (a *Agent) consulConfig() (*consul.Config, error) {
 
 	base.RPCAddr = a.config.RPCBindAddr
 	base.RPCAdvertise = a.config.RPCAdvertiseAddr
-
-	if a.config.ReconnectTimeoutLAN != 0 {
-		base.SerfLANConfig.ReconnectTimeout = a.config.ReconnectTimeoutLAN
-	}
-	if a.config.ReconnectTimeoutWAN != 0 {
-		base.SerfWANConfig.ReconnectTimeout = a.config.ReconnectTimeoutWAN
-	}
 
 	base.Segment = a.config.SegmentName
 	if len(a.config.Segments) > 0 {

--- a/agent/agent_test.go
+++ b/agent/agent_test.go
@@ -196,6 +196,19 @@ func TestAgent_ReconnectConfigSettings(t *testing.T) {
 	}()
 }
 
+func TestAgent_ReconnectConfigWanDisabled(t *testing.T) {
+	t.Parallel()
+
+	a := NewTestAgent(t.Name(), `
+		ports { serf_wan = -1 }
+		reconnect_timeout_wan = "36h"
+	`)
+	defer a.Shutdown()
+
+	// This is also testing that we dont panic like before #4515
+	require.Nil(t, a.consulConfig().SerfWANConfig)
+}
+
 func TestAgent_setupNodeID(t *testing.T) {
 	t.Parallel()
 	a := NewTestAgent(t.Name(), `

--- a/agent/testagent.go
+++ b/agent/testagent.go
@@ -127,9 +127,9 @@ func (a *TestAgent) Start() *TestAgent {
 
 	for i := 10; i >= 0; i-- {
 		a.Config = TestConfig(
+			randomPortsSource(a.UseTLS),
 			config.Source{Name: a.Name, Format: "hcl", Data: a.HCL},
 			config.Source{Name: a.Name + ".data_dir", Format: "hcl", Data: hclDataDir},
-			randomPortsSource(a.UseTLS),
 		)
 
 		// write the keyring

